ports/zephyr/zephyr_device.c

```diff
@@ -47,3 +122,94 @@ const struct device *zephyr_device_find(mp_obj_t name) {
+
+const struct device *zephyr_device_setup(mp_obj_t obj) {
+    mp_obj_t *args;
+    mp_obj_t *pinctrls;
+    const mp_zephyr_device_data_t *device;
+    size_t pinctrl_count;
+    const char *dev_name;
+    pinctrl_soc_pin_t *pinctrl;
+    struct pinctrl_state *pinctrl_states;
+    int ret = 0;
+
+    if (mp_obj_is_str(obj)) {
+        return zephyr_device_find(obj);
+    }
+    if (mp_obj_get_int(mp_obj_len(obj)) != 2) {
+        mp_raise_ValueError(MP_ERROR_TEXT("Invalid device object"));
+    }
+
+    mp_obj_get_array_fixed_n(obj, 2, &args);
+
+    dev_name = mp_obj_str_get_str(args[0]);
+    device = device_find_by_node_label(dev_name);
+
+    if (device == NULL) {
+        device = device_find_by_name(dev_name);
+    }
+
+    if (device == NULL) {
+        #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
+        mp_raise_ValueError(MP_ERROR_TEXT("device not found"));
+        #else
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("device %s not found"), dev_name);
+        #endif
+    } else if (device->dev == NULL || device->pinctrl_cfg == NULL) {
+        #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
+        mp_raise_ValueError(MP_ERROR_TEXT("device is invalid"));
+        #else
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("device %s is invalid"), dev_name);
+        #endif
+    }
+
+    if (device_is_ready(device->dev)) {
+        ret = device_deinit(device->dev);
+    }
+
+    if (ret == -ENOTSUP) {
+        mp_raise_msg(&mp_type_RuntimeError,
+            MP_ERROR_TEXT("Device is already initialized and doesn't support de-initialization"));
+    }
+    if (ret < 0) {
+        mp_raise_OSError(-ret);
+    }
+
+    if (mp_obj_get_int(mp_obj_len(args[1])) < 1) {
+        mp_raise_ValueError(MP_ERROR_TEXT("No Pinctrl objects"));
+    }
+    mp_obj_get_array(args[1], &pinctrl_count, &pinctrls);
+
+    /* should get cleaned by gc properly when pinctrl dont need it anymore? Otherwise, leaks */
+    pinctrl = m_malloc(pinctrl_count * sizeof(pinctrl_soc_pin_t));
+    pinctrl_states = m_malloc(device->pinctrl_cfg->state_cnt * sizeof(struct pinctrl_state));
+    for (int i = 0; i < pinctrl_count; i++) {
+        pinctrl[i] = zephyr_pinctrl_get_pinctrl(pinctrls[i]);
+    }
+    for (int i = 0; i < device->pinctrl_cfg->state_cnt; i++) {
+        pinctrl_states[i].pins = pinctrl;
+        pinctrl_states[i].pin_cnt = pinctrl_count;
+        pinctrl_states[i].id = device->pinctrl_cfg->states[i].id;
+    }
+    ret = pinctrl_update_states(device->pinctrl_cfg, pinctrl_states, device->pinctrl_cfg->state_cnt);
+    if (ret < 0) {
+        mp_raise_msg(&mp_type_RuntimeError, MP_ERROR_TEXT("Couldn't apply pinctrl"));
+    }
+
+    ret = device_init(device->dev);
+    if (ret < 0) {
+        mp_raise_OSError(-ret);
+    }
+
+    return device->dev;
+}
+
+#else
+
+const struct device *zephyr_device_setup(mp_obj_t obj) {
+    return zephyr_device_find(obj);
+}
```

Comment:
For boards not yet supported, nothing is broken: new board contributed can add the feature in an incremental fashion.